modules/dop/services/issue/validate.go

```diff
@@ -0,0 +1,107 @@
+// Copyright (c) 2021 Terminus, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package issue
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/erda-project/erda/apistructs"
+	"github.com/erda-project/erda/modules/dop/dao"
+)
+
+type issueValidator struct {
+	iteration *dao.Iteration
+	state     *dao.IssueState
+	db        *dao.DBClient
+}
+
+type issueValidationConfig struct {
+	iterationID int64
+	stateID     int64
+}
+
+func NewIssueValidator(c *issueValidationConfig, db *dao.DBClient) (*issueValidator, error) {
+	v := issueValidator{db: db}
+	if c == nil {
+		return &v, nil
+	}
+	if _, err := v.TryGetIteration(c.iterationID); err != nil {
+		return nil, err
+	}
+	if _, err := v.TryGetState(c.stateID); err != nil {
+		return nil, err
+	}
+	return &v, nil
+}
+
+func (v *issueValidator) TryGetIteration(iterationID int64) (*dao.Iteration, error) {
+	if iterationID <= 0 {
+		return nil, nil
+	}
+	if v.iteration != nil {
+		return v.iteration, nil
+	}
+	iteration, err := v.db.GetIteration(uint64(iterationID))
+	if err != nil {
+		return nil, fmt.Errorf("failed to get iteration, err: %v", err)
```

Comment:
as a `cache`, if it error every time, should we cache the errors?

Reply:
No, we cache value instead of err